common/path.cpp

```diff
@@ -20,67 +20,119 @@
  */
 
 #include "common/path.h"
+#include "common/punycode.h"
+#include "common/hash-str.h"
+
+const char ESCAPER = '/';
+const char ESCAPE_SLASH = '+';
+const char ESCAPE_SEPARATOR = '/';
+const char *DIR_SEPARATOR = "//";
+const char *SLASH_ESCAPED = "/+";
 
 namespace Common {
 
 Path::Path(const Path &path) {
-	_str = path.rawString();
+	_str = path._str;
 }
 
 Path::Path(const char *str, char separator) {
 	set(str, separator);
 }
 
 Path::Path(const String &str, char separator) {
-	if (separator == DIR_SEPARATOR)
-		_str = str;
-	else
-		set(str.c_str(), separator);
+	set(str.c_str(), separator);
 }
 
 String Path::toString(char separator) const {
 	String res;
-	for (const char *ptr = _str.c_str(); *ptr; ptr++) {
-		if (*ptr == DIR_SEPARATOR)
-			res += separator;
-		else
-			res += *ptr;
+	for (uint i = 0; i < _str.size(); i++) {
+		if (_str[i] == ESCAPER) {
+			i++;
+			if (_str[i] == ESCAPE_SLASH)
+				res += '/';
+			else if (_str[i] == ESCAPE_SEPARATOR)
+				res += separator;
+		} else
+			res += _str[i];
 	}
 	return res;
 }
 
+size_t Path::findLastSeparator() const {
+	size_t res = Common::String::npos;
+	for (uint i = 0; i + 2 < _str.size(); i++) {
+		if (_str[i] == ESCAPER) {
+			i++;
+			if (_str[i] == ESCAPE_SEPARATOR)
+				res = i - 1;
+		}
+	}
+
+	return res;
+}
+
 Path Path::getParent() const {
 	if (_str.size() < 2)
 		return Path();
-	size_t separatorPos = _str.findLastOf(DIR_SEPARATOR, _str.size() - 2);
+	size_t separatorPos = findLastSeparator();
 	if (separatorPos == Common::String::npos)
 		return Path();
-	return Path(_str.substr(0, separatorPos + 1), DIR_SEPARATOR);
+	Path ret;
+	ret._str = _str.substr(0, separatorPos + 2);
+	return ret;
 }
 
 Path Path::getLastComponent() const {
 	if (_str.size() < 2)
 		return *this;
-	size_t separatorPos = _str.findLastOf(DIR_SEPARATOR, _str.size() - 2);
+	size_t separatorPos = findLastSeparator();
 	if (separatorPos == Common::String::npos)
 		return *this;
-	return Path(_str.substr(separatorPos + 1), DIR_SEPARATOR);
+	Path ret;
+	ret._str = _str.substr(separatorPos + 2);
+	return ret;
+}
+
+static String escapePath(const String& in) {
+	String ret;
+	for (uint i = 0; i < in.size(); i++) {
+		if (in[i] == '/')
+			ret += SLASH_ESCAPED;
+		else
+			ret += in[i];
+	}
+	return ret;
+}
+
+Path Path::appendComponent(const String &x) const {
+	if (x.empty())
+		return *this;
+	Common::String str = _str;
```

Comment:
You're in namespace Common, may write just `String`

Reply:
Done